File: src/lib/defcon.ts
import type { PositionCode } from '@/lib/types'

// DefConPlayer shape — canonical definition lives here until Plan 01 adds it to types.ts.
// When types.ts gains DefConPlayer, update this import to: import type { DefConPlayer, PositionCode } from '@/lib/types'
export interface DefConPlayer {
  id: number
  web_name: string
  element_type: PositionCode       // 2=DEF, 3=MID, 4=FWD
  team: number
  team_short_name: string
  threshold: number                // 10 for DEF, 12 for MID/FWD
  hit_rate: number                 // 0.0-1.0
  hits: number
  games_played: number
  avg_per90: number
  distance_to_threshold: number    // threshold - avg_per90 (negative = above threshold)
  fixture_correlation: {
    insufficient_data: boolean
    easy_hit_rate?: number
    hard_hit_rate?: number
    easy_n?: number
    hard_n?: number
  }
}

// Per-position DefCon thresholds: DEF needs 10, MID/FWD need 12
export const DEFCON_THRESHOLD: Record<number, number> = {
  2: 10,  // DEF
  3: 12,  // MID
  4: 12,  // FWD
}

// Splits players into DEF (element_type=2) and MID/FWD (element_type=3 or 4) groups
export function splitByPosition(players: DefConPlayer[]): {
  def: DefConPlayer[]
  midFwd: DefConPlayer[]
} {
  return {
    def: players.filter(p => p.element_type === 2),
    midFwd: players.filter(p => p.element_type === 3 || p.element_type === 4),
  }
}

// Formats a 0.0-1.0 hit rate as a percentage string, e.g. 0.516 -> "51.6%"
export function formatHitRate(rate: number): string {
  return `${(rate * 100).toFixed(1)}%`
}

// Returns the player's DefCon status relative to their threshold
export function getDefConStatus(player: DefConPlayer): 'above' | 'at' | 'below' {
  if (player.distance_to_threshold < 0) return 'above'
  if (player.distance_to_threshold === 0) return 'at'
  return 'below'
}

// Formats a fixture correlation result for display
export function formatCorrelation(fc: DefConPlayer['fixture_correlation']): {
  label?: string
  easy?: string
  hard?: string
} {
  if (fc.insufficient_data) {
    return { label: `Insufficient data (${fc.easy_n ?? 0} easy, ${fc.hard_n ?? 0} hard games)` }
  }
  return {
    easy: `${((fc.easy_hit_rate ?? 0) * 100).toFixed(1)}%`,
    hard: `${((fc.hard_hit_rate ?? 0) * 100).toFixed(1)}%`,
  }
}

File: pipeline/defcon.py
"""Compute DefCon stats from FPL element-summary per-match history."""

DEFCON_THRESHOLD = {2: 10, 3: 12, 4: 12}  # position_code -> threshold


def compute_defcon_stats(bootstrap: dict, difficulty_scores: dict, summaries: dict) -> list:
    """
    For each DEF/MID/FWD player with starts > 0, look up element-summary from
    the pre-fetched summaries dict and compute hit rate, avg per90, distance to
    threshold, and fixture correlation.

    Args:
        bootstrap: Full FPL bootstrap-static JSON
        difficulty_scores: dict mapping team_id (int) -> difficulty score (0.0-1.0),
                          computed by merge.py from rolling xGA
        summaries: dict mapping player_id (int) -> element-summary response dict.
                   Pre-fetched by run.py shared cache.
    Returns:
        List of dicts matching DefConPlayer interface shape
    """
    teams = {t['id']: t for t in bootstrap.get('teams', [])}
    results = []

    for element in bootstrap['elements']:
        pos = element['element_type']
        if pos not in (2, 3, 4):
            continue
        if element.get('starts', 0) == 0:
            continue

        threshold = DEFCON_THRESHOLD[pos]

        summary = summaries.get(element['id'])
        if summary is None:
            continue

        history = [m for m in summary.get('history', []) if m['minutes'] > 0]
        games_played = len(history)
        if games_played < 5:
            continue

        hits = sum(1 for m in history if m.get('defensive_contribution', 0) >= threshold)
        hit_rate = round(hits / games_played, 4)

        avg_per90 = element.get('defensive_contribution_per_90') or 0.0
        distance = round(threshold - avg_per90, 2)

        # DEF-03: fixture difficulty correlation
        fixture_correlation = _compute_fixture_correlation(history, difficulty_scores, threshold)

        team_id = element['team']
        results.append({
            'id': element['id'],
            'web_name': element['web_name'],
            'element_type': pos,
            'team': team_id,
            'team_short_name': teams.get(team_id, {}).get('short_name', ''),
            'threshold': threshold,
            'hit_rate': hit_rate,
            'hits': hits,
            'games_played': games_played,
            'avg_per90': round(float(avg_per90), 2),
            'distance_to_threshold': distance,
            'fixture_correlation': fixture_correlation,
        })

    return results


def _compute_fixture_correlation(history: list, difficulty_scores: dict, threshold: int) -> dict:
    """Split games into easy vs hard fixtures and compare hit rates.

    Easy = opponent difficulty_score < 0.4 (opponent concedes many goals)
    Hard = opponent difficulty_score > 0.6 (opponent concedes few goals)
    Returns insufficient_data: true when either bucket has < 5 games.
    """
    easy_games = [m for m in history if difficulty_scores.get(m.get('opponent_team'), 0.5) < 0.4]
    hard_games = [m for m in history if difficulty_scores.get(m.get('opponent_team'), 0.5) > 0.6]

    if len(easy_games) < 5 or len(hard_games) < 5:
        return {
            'insufficient_data': True,
            'easy_n': len(easy_games),
            'hard_n': len(hard_games),
        }

    easy_hits = sum(1 for m in easy_games if m.get('defensive_contribution', 0) >= threshold)
    hard_hits = sum(1 for m in hard_games if m.get('defensive_contribution', 0) >= threshold)

    return {
        'insufficient_data': False,
        'easy_hit_rate': round(easy_hits / len(easy_games), 4),
        'hard_hit_rate': round(hard_hits / len(hard_games), 4),
        'easy_n': len(easy_games),
        'hard_n': len(hard_games),
    }
